ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Genesis/Governor.hs

```diff
@@ -162,52 +214,83 @@ densityDisconnect ::
-            lastSlotForecast > endOfGenesisWindow
-          hasBlockAfterGenesisWindow =
-            candidateHasBlockAfter || forecastHasBlockAfter
-          upperBound =
-              lowerBound
-            + if hasBlockAfterGenesisWindow
-              then 0
-              else unSlotNo (intervalLength unresolvedSlotsLB unresolvedSlotsUB)
-          offersMoreThanK
-            | candidateHasBlockAfter
-            = AF.length candidateSuffix > fromIntegral k
+
+          caughtUp = Set.member peer caughtUpPeers
+
+          latestSlot = case (AF.headSlot candidateSuffix, latestSlots Map.!? peer) of
+            (Origin, Nothing) -> NoLatestSlot
+            (Origin, Just latest)
+              | latest > 0 -> LatestSlotForecast latest
+              | otherwise -> NoLatestSlot
+            (NotOrigin cand, Nothing) -> LatestSlotCandidate cand
+            (NotOrigin cand, Just latest)
+              | latest > cand -> LatestSlotForecast latest
+              | otherwise -> LatestSlotCandidate cand
+
+          -- If the peer has more headers that it hasn't sent yet, each slot between the latest header we know of and
+          -- the end of the Genesis window could contain a block, so the upper bound for the total number of blocks in
+          -- the window is the sum of the known blocks and that number of remaining slots.
+          -- If the slot of the latest header we know of is _after_ the end of the Genesis window (either because the
+          -- candidate fragment extends beyond it or because we are waiting to validate a header beyond the forecast
+          -- horizon that we already received), there can be no headers in between and 'potentialSlots' is 0.
+          SlotNo potentialSlots
+            | caughtUp
+            = 0
             | otherwise
-            = AF.length candidateSuffix >= fromIntegral k
-          lastSlot | isJust maybeForecast, lastSlotForecast >= lastSlotCandidate = Left lastSlotForecast
-                   | otherwise = Right lastSlotCandidate
-      pure (peer, DensityBounds {fragment, offersMoreThanK, lowerBound, upperBound, hasBlockAfter = hasBlockAfterGenesisWindow, lastSlot})
+            = case latestSlot of
+              -- The peer either has not advertised its tip yet or simply has no blocks whatsoever and won't progress
+              -- either.
+              -- In the latter case, it should be killed by the LoP.
+              -- REVIEW: What do we want to do in tests here? Both cases are possible, and we cannot distinguish them
+              -- without the tip (even with the tip we have been relying on the invariant that each peer advertises the
+              -- tip before any of them sends headers).
+              -- NoLatestSlot -> 0
+              NoLatestSlot -> SlotNo sgen
+              -- 'endOfGenesisWindow' is exclusive, so we have to add 1 to the non-exclusive last slot
+              LatestSlotCandidate slot -> intervalLength endOfGenesisWindow (slot + 1)
```

Comment:
I think we want instead `intervalLength (slot + 1) endOfGenesisWindow`.

Reply:
damn.